pkg/controller/operatingsystemconfig/actuator.go

```diff
@@ -0,0 +1,223 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operatingsystemconfig
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"path/filepath"
+
+	"github.com/gardener/gardener/extensions/pkg/controller/operatingsystemconfig"
+	oscommonactuator "github.com/gardener/gardener/extensions/pkg/controller/operatingsystemconfig/oscommon/actuator"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/gardener/gardener-extension-os-gardenlinux/pkg/controller/operatingsystemconfig/generator"
+	"github.com/gardener/gardener-extension-os-gardenlinux/pkg/gardenlinux"
+	"github.com/gardener/gardener-extension-os-gardenlinux/pkg/memoryone"
+)
+
+type actuator struct {
+	client               client.Client
+	useGardenerNodeAgent bool
+}
+
+// NewActuator creates a new Actuator that updates the status of the handled OperatingSystemConfig resources.
+func NewActuator(mgr manager.Manager, useGardenerNodeAgent bool) operatingsystemconfig.Actuator {
+	return &actuator{
+		client:               mgr.GetClient(),
+		useGardenerNodeAgent: useGardenerNodeAgent,
+	}
+}
+
+func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, osc *extensionsv1alpha1.OperatingSystemConfig) ([]byte, *string, []string, []string, []extensionsv1alpha1.Unit, []extensionsv1alpha1.File, error) {
+	cloudConfig, command, err := oscommonactuator.CloudConfigFromOperatingSystemConfig(ctx, log, a.client, osc, generator.CloudInitGenerator())
+	if err != nil {
+		return nil, nil, nil, nil, nil, nil, fmt.Errorf("could not generate cloud config: %w", err)
+	}
+
+	switch purpose := osc.Spec.Purpose; purpose {
+	case extensionsv1alpha1.OperatingSystemConfigPurposeProvision:
+		if !a.useGardenerNodeAgent {
+			return cloudConfig, command, oscommonactuator.OperatingSystemConfigUnitNames(osc), oscommonactuator.OperatingSystemConfigFilePaths(osc), nil, nil, nil
+		}
+		userData, err := a.handleProvisionOSC(ctx, osc)
+		return []byte(userData), nil, nil, nil, nil, nil, err
+
+	case extensionsv1alpha1.OperatingSystemConfigPurposeReconcile:
+		extensionUnits, extensionFiles, err := a.handleReconcileOSC(osc)
+		return cloudConfig, command, oscommonactuator.OperatingSystemConfigUnitNames(osc), oscommonactuator.OperatingSystemConfigFilePaths(osc), extensionUnits, extensionFiles, err
```

Comment:
Could you add a comment here that after the migration to gardener-node-agent is done & gardener-node-agent feature gate is removed in the gardenlet, we only need to return the extra units and files? (then we can do `return nil, nil, nil, nil, extensionUnits, extensionFiles, err`)